### PR TITLE
CXX-506 CXX-431 CXX-432 CXX-434 Use MNMLSTC for 14 and TS polyfills

### PR DESCRIPTION
**Pull request by acmorrow**

### CODE REVIEW

File: src/bsoncxx/CMakeLists.txt [hunk 1]
```diff
@@ -23,82 +48,32 @@ include_directories(
...
-    ${bsoncxx-private_sources}
-    ${bsoncxx-stdx_sources}
-    ${bsoncxx-util_sources}
+add_library(bsoncxx_static STATIC
+  ${bsoncxx_sources}
 )
 
-add_library(bsoncxx SHARED ${bsoncxx_sources})
+set_target_properties(bsoncxx_static
+  PROPERTIES OUTPUT_NAME bson
```
Comment (TylerBrock) [inline src/bsoncxx/CMakeLists.txt]: PROPERTIES should be on the line above

File: src/bsoncxx/CMakeLists.txt [hunk 2]
```diff
@@ -1,21 +1,46 @@
...
+set(BSONCXX_VERSION ${BSONCXX_VERSION_MAJOR}.${BSONCXX_VERSION_MINOR}.${BSONCXX_VERSION_PATCH})
+set(BSONCXX_INLINE_NAMESPACE "v${BSONCXX_ABI_VERSION}")
 set(BSONCXX_DIRECTORY_PREFIX "v${BSONCXX_VERSION_MAJOR}.${BSONCXX_VERSION_MINOR}")
-set(BSONCXX_INSTALL_DIR "include/bsoncxx/${BSONCXX_DIRECTORY_PREFIX}")
+set(BSONCXX_HEADER_INSTALL_DIR "include/bsoncxx/${BSONCXX_DIRECTORY_PREFIX}" CACHE INTERNAL "")
+
+# TODO: Not all combinations of options are valid, and USE_STD is only valid
+# for C++14 compiles. Add logic to validate options and configure checks to validate
+# STD_EX vs C++14.
+option (BSONCXX_POLY_USE_STD_EXPERIMENTAL "Use <experimental/ for stdx polyfills" OFF)
```
Comment (TylerBrock) [inline src/bsoncxx/CMakeLists.txt]: no space here between option and args

File: src/bsoncxx/CMakeLists.txt [hunk 3]
```diff
@@ -1,21 +1,46 @@
...
+option (BSONCXX_POLY_USE_SYSTEM_MNMLSTC "Obtain mnmlstc/core from system" OFF)
 
-set(LIBBSON_REQUIRED_ABI_VERSION 1.0)
 pkg_check_modules(LIBBSON REQUIRED libbson-${LIBBSON_REQUIRED_ABI_VERSION})
 
-link_directories(${LIBBSON_LIBRARY_DIRS})
+add_subdirectory(third_party)
+add_subdirectory(config)
+
+set (bsoncxx_sources
```
Comment (TylerBrock) [inline src/bsoncxx/CMakeLists.txt]: no space between set and sources

File: src/mongocxx/test/CMakeLists.txt [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
+  write_concern.cpp
+)
 
 include_directories(
     ${CMAKE_SOURCE_DIR}/src/third_party/catch/include
-    ${LIBBSON_INCLUDE_DIRS}
-    ${LIBMONGOC_INCLUDE_DIRS}
-    ${CMAKE_SOURCE_DIR}/src
-    ${CMAKE_BINARY_DIR}/src
+#    ${LIBBSON_INCLUDE_DIRS}
```
Comment (TylerBrock) [inline src/mongocxx/test/CMakeLists.txt]: Can we nuke these lines if they are commented out?

**Comment by TylerBrock** [comment]
I'd like to try to use 4 space everywhere.

**Comment by amidvidy** [comment]
this LGTM if travis is happy.

**Comment by hanumantmk** [comment]
LGTM

File: src/bsoncxx/CMakeLists.txt (hunk 2 above)
Comment (acmorrow) [inline src/bsoncxx/CMakeLists.txt]: Done

File: src/bsoncxx/CMakeLists.txt (hunk 3 above)
Comment (acmorrow) [inline src/bsoncxx/CMakeLists.txt]: Done

File: src/bsoncxx/CMakeLists.txt (hunk 1 above)
Comment (acmorrow) [inline src/bsoncxx/CMakeLists.txt]: Done

File: src/mongocxx/test/CMakeLists.txt (hunk 4 above)
Comment (acmorrow) [inline src/mongocxx/test/CMakeLists.txt]: Done

**Comment by acmorrow** [comment]
I've tried to fix up the 2 vs 4 space indent as well.

**Pull request merged**